### ISSUE TITLE
AMQP illegal-state sender is closed

### ISSUE THREAD

**Issue by backspacer80** [Azure/azure-iot-sdk-csharp#1208]
- **OS, version, SKU and CPU architecture used:** Windows Server 2016 Standard Version 1604 x64
- **Application's .NET Target Framework :** netcoreapp3.1, netcoreapp2.1 (occurs on both)
- **Device:** ESXi VM
- **SDK version used:** Microsoft.Azure.Devices.Client 1.21.1

## Description of the issue:
We have an ASP.NET Core application that plays the role of an IoT gateway. It collects data from controllers and sends them to IoT hub. There are 2-8 messages that are sent out every 5 seconds using the `public Task SendEventAsync(Message message)` method of Azure Device Client. After one or two days of stable operation communication breaks down with the exception below and never recovers itself. When the application keeps running then sender is closed exception pops up with multiple variations of senderXYZ for almost all messages. However, some messages still become delivered successfully.

## Code sample exhibiting the issue:
```
var message = new Message(MessageUtils.MessageEncoding.GetBytes(messageString));
var deviceClient = GetDeviceClient(gatewayDeviceId, deviceId);
deviceClient.SendEventAsync(message).ContinueWith(OnSendHighPriorityEventCompleted, message);

// DeviceClient instances are created once and then reused. This is how they are created inside GetDeviceClient:

            var auth = new DeviceAuthenticationWithRegistrySymmetricKey(registeredDevice.Id, registeredDevice.Token);

            deviceClient = DeviceClient.Create(
                hostName,
                auth,
                new ITransportSettings[]
                {
                    new AmqpTransportSettings(AzureConstants.TRANSPORT_TYPE)
                    {
                        //  Setting Pooling to true enables connection multiplexing between devices.
                        AmqpConnectionPoolSettings = new AmqpConnectionPoolSettings()
                        {
                            Pooling = true,
                            MaxPoolSize = 1
                        }
                    }
                }
            );

            // Use our own retry policy because we have requirement to persist unsent messages
            deviceClient.SetRetryPolicy(new NoRetry());
            deviceClient.OperationTimeoutInMilliseconds = 60000;
        }

```
## Console log of the issue:
System.AggregateException: One or more errors occurred. (error(condition:amqp:illegal-state,description:'sender185' is closed))
 ---> Microsoft.Azure.Devices.Client.Exceptions.IotHubCommunicationException: error(condition:amqp:illegal-state,description:'sender185' is closed)
 ---> Microsoft.Azure.Devices.Client.Exceptions.IotHubException: error(condition:amqp:illegal-state,description:'sender185' is closed)
 ---> Microsoft.Azure.Amqp.AmqpException: 'sender185' is closed
   at Microsoft.Azure.Amqp.AmqpObject.ThrowIfClosed()
   at Microsoft.Azure.Amqp.SendingAmqpLink.BeginSendMessage(AmqpMessage message, ArraySegment`1 deliveryTag, ArraySegment`1 txnId, TimeSpan timeout, AsyncCallback callback, Object state)
   at Microsoft.Azure.Amqp.SendingAmqpLink.<>c__DisplayClass11_0.<SendMessageAsync>b__0(AsyncCallback c, Object s)
   at System.Threading.Tasks.TaskFactory`1.FromAsyncImpl(Func`3 beginMethod, Func`2 endFunction, Action`1 endAction, Object state, TaskCreationOptions creationOptions)
   at Microsoft.Azure.Amqp.SendingAmqpLink.SendMessageAsync(AmqpMessage message, ArraySegment`1 deliveryTag, ArraySegment`1 txnId, TimeSpan timeout)
   at Microsoft.Azure.Devices.Client.Transport.AmqpIoT.AmqpIoTSendingLink.SendAmqpMessageAsync(AmqpMessage amqpMessage, TimeSpan timeout)
   --- End of inner exception stack trace ---
   --- End of inner exception stack trace ---
   at Microsoft.Azure.Devices.Client.Transport.AmqpIoT.AmqpIoTSendingLink.SendAmqpMessageAsync(AmqpMessage amqpMessage, TimeSpan timeout)
   at Microsoft.Azure.Devices.Client.Transport.AmqpIoT.AmqpIoTSendingLink.SendMessageAsync(Message message, TimeSpan timeout)
   at Microsoft.Azure.Devices.Client.Transport.AmqpIoT.AmqpUnit.SendMessageAsync(Message message, TimeSpan timeout)
   at Microsoft.Azure.Devices.Client.Transport.AmqpIoT.AmqpUnit.SendEventAsync(Message message, TimeSpan timeout)
   at Microsoft.Azure.Devices.Client.Transport.Amqp.AmqpTransportHandler.SendEventAsync(Message message, CancellationToken cancellationToken)
   at Microsoft.Azure.Devices.Client.Transport.ErrorDelegatingHandler.<>c__DisplayClass22_0.<<ExecuteWithErrorHandlingAsync>b__0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.Azure.Devices.Client.Transport.ErrorDelegatingHandler.ExecuteWithErrorHandlingAsync[T](Func`1 asyncOperation)
   at Microsoft.Azure.Devices.Client.Transport.RetryDelegatingHandler.<>c__DisplayClass14_0.<<SendEventAsync>b__0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.Azure.Devices.Client.Transport.RetryDelegatingHandler.SendEventAsync(Message message, CancellationToken cancellationToken)
   at Microsoft.Azure.Devices.Client.InternalClient.SendEventAsync(Message message)
   --- End of inner exception stack trace ---

**Comment by timtay-microsoft** [comment]
There have been some fixes made to the AMQP layer of the device client since version 1.21.1. Do you see the same problem after upgrading to the latest version?

**Comment by backspacer80** [comment]
Hi @timtay-microsoft , I've upgraded to 1.22.0 and the application has been running for a week without issues. However, I also have a version running on the same machine that still uses 1.21.1, also without issues. There were no other modifications from my side. So whether that's an evidence that the issue was fixed in version 1.22.0 or not I don't know. But for now I can set the issue to closed.

**Comment by az-iot-builder-01** [comment]
@backspacer80, thank you for your contribution to our open-sourced project! Please help us improve by filling out this 2-minute [customer satisfaction survey](https://forms.office.com/Pages/ResponsePage.aspx?id=v4j5cvGGr0GRqy180BHbRx8jpG79OudPqY-BojcKjkJUMFFBQjgxNkw5ODJIRjBWTzI0Nk85R0pUQiQlQCN0PWcu)
